maddymz/irondb
# US303

Adds /database/export route where the user can export the selected data to csv. Current there is an issue, most likely related to local host where the file is renamed to strictly "export" when downloaded. However the file being written is correctly named. This will be a future task. Also note the UI/options are limited as this US covers basic functionality only.

Video of expected behavior (watch on 2x speed): https://www.youtube.com/watch?v=36mYuAT25T0&feature=youtu.be

Before testing: Few things about this pull request. For csv creation pr adds json2csv, since this is a new npm dependency you will most likely need to rebuild your docker images. For this reason it is probably best to run on a clean git clone.



To test:
1. Basic setup: After cloning, navigating to repo, and checking out to branch `US303` run: `npm install`, `gulp sass`, `gulp js`, `mkdir pg-data`
2. Check that there is a directory name `temp` in the `/public` directory, i.e. `irondb/public/temp`. There should be a placeholder.md in the directory. This is where files will be written to and then deleted once sent.
3. Remove images and containers, run: `docker rm $(docker ps -a -q)` then `docker rmi $(docker images -q)`
4. Build and deploy with `docker-compose up --build` (note: this will take some time)
Database initialization will also take a bit of time, just considerably less than docker.
5. Once everything is running confirm server is up by going to `localhost:3001` in your browser.
6. See video above for in-depth expected behavior: 
      Summary (zoomed out for overview):
     -  get request on /database/export outputs: 
![screen shot 2018-11-29 at 8 44 40 am](https://user-images.githubusercontent.com/6512755/49225940-4c52f200-f3b3-11e8-90de-6e1a324bbab3.jpg)
     - Button on /database page on left bottom of search panel (note: styling will change once bootstrap customized, i.e. more visible):
![screen shot 2018-11-29 at 8 45 39 am](https://user-images.githubusercontent.com/6512755/49225985-67bdfd00-f3b3-11e8-9263-a56183a0d6b7.jpg)
      - Clicking red x-circles remove given row, if all row removed output, not attempting to export no data outputs a 400 error:
![screen shot 2018-11-29 at 8 45 24 am](https://user-images.githubusercontent.com/6512755/49226083-a5bb2100-f3b3-11e8-8df4-06c2016ecee4.jpg)
     - Successful data export:
![screen shot 2018-11-29 at 8 50 21 am](https://user-images.githubusercontent.com/6512755/49226246-12362000-f3b4-11e8-8fce-40913b40b7ea.jpg)
![screen shot 2018-11-29 at 8 52 38 am](https://user-images.githubusercontent.com/6512755/49226288-2aa63a80-f3b4-11e8-8551-033379605f6b.jpg)
    - Adding file extension .csv to export it will open correctly formatted in excel, pages, etc. Also check that the temp directory only contains the placeholder after export.
    - Output in chrome:
![screen shot 2018-11-29 at 8 55 13 am](https://user-images.githubusercontent.com/6512755/49226482-bae47f80-f3b4-11e8-8bc2-f7c031a074ab.jpg)
      - Finally note: safari opens data, but not file exactly.
